src/Controllers/DatabaseController.js

```diff
@@ -59,28 +59,37 @@ DatabaseController.prototype.validateClassName = function(className) {
   return Promise.resolve();
 };
 
-// Returns a promise for a schema object.
-// If we are provided a acceptor, then we run it on the schema.
-// If the schema isn't accepted, we reload it at most once.
-DatabaseController.prototype.loadSchema = function(acceptor = returnsTrue) {
+function _loadSchema(db) {
+  if (db._schema) {
+    return Promise.resolve(db._schema);
+  }
 
-  if (!this.schemaPromise) {
-    this.schemaPromise = this.schemaCollection().then(collection => {
-      delete this.schemaPromise;
+  if (!db.schemaPromise) {
+    db.schemaPromise = db.schemaCollection().then(collection => {
       return Schema.load(collection);
-    });
-    return this.schemaPromise;
+    }).then((res) => {
+      db._schema = res;
+      return Promise.resolve(db._schema);
+    })
   }
+  return db.schemaPromise;
+}
 
-  return this.schemaPromise.then((schema) => {
+DatabaseController.prototype.invalidateSchema = function() {
+  delete this._schema;
+  delete this.schemaPromise;
```

Comment:
You would still need to ``delete this.schemaPromise`` in the ``_loadSchema`` function similar to how it was originally

Compare these two scenarios where the server gets two requests (``R1``,``R2``) in quick succession, Say a initial page load that does couple queries

With the setup in this pull request
```javascript
R1.invalidateSchema
R1._loadSchema
R1.schemaCollection // creates schemaPromise (Queries _SCHEMA)
R2.invalidateSchema // deletes schemaPromise before it has even returned
R2._loadSchema
R2.schemaCollection // creates schemaPromise (Queries _SCHEMA)
// R1.schemaCollection query resolves
// R2.schemaCollection query resolves

// This would result in two concurrent _SCHEMA queries
```


If you moved the ``delete this.scehamPromise`` into the ``_loadSchema``

```javascript
R1.invalidateSchema
R1._loadSchema
R1.schemaCollection // creates schemaPromise (Queries _SCHEMA)
R2.invalidateSchema // doesnt delete schemaPromise
R2._loadSchema
// R1.schemaCollection resolves
// R1 deletes schemaPromise
// R2 resolves from R1.schemaCollection resolution

// This would result in only one _SCHEMA query
```